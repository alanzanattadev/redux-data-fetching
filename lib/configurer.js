"use babel";
// @flow
import { schema } from "normalizr";
import { buildSchema } from "graphql";
import type { GraphQLSchema } from "graphql";
import configureMiddleware from "./middleware";
import configureReducer from "./reducer";
import configureActions from "./actions";
import configureConnecter from "./hoc";
import { graphQLizr, graphQLRecordr } from "./graphqlTypesConverters";
import { convertsTypesSchemaToSelectorSchema } from "./selectors";

export function configure(
  graphQLSchema: GraphQLSchema | string,
  context: mixed,
) {
  const graphQLCompiledSchema =
    typeof graphQLSchema === "string"
      ? buildSchema(graphQLSchema)
      : graphQLSchema;
  const selectorSchema = convertsTypesSchemaToSelectorSchema(
    graphQLCompiledSchema,
  );
  const normalizrModel = graphQLizr(graphQLCompiledSchema);
  const recordsModel = graphQLRecordr(graphQLCompiledSchema);
  const actions = configureActions();
  const middleware = configureMiddleware(
    graphQLCompiledSchema,
    actions,
    normalizrModel,
    context,
  );
  const reducer = configureReducer(
    normalizrModel.entities,
    recordsModel,
    graphQLCompiledSchema,
  );
  const GraphQLConnecter = configureConnecter({ selectorSchema });
  return {
    actions,
    middleware,
    reducer,
    normalizrModel,
    recordsModel,
    selectorSchema,
    GraphQLConnecter,
  };
}
