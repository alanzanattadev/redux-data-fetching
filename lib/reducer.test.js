"use babel";
// @flow

import React from "react";
import { shallow, mount } from "enzyme";
import { configure } from "./configurer";
import { Map, fromJS } from "immutable";
import {
  printSchema,
  graphql,
  GraphQLObjectType,
  GraphQLSchema,
  GraphQLString,
  GraphQLFloat,
  GraphQLList,
  GraphQLInt,
  GraphQLBoolean,
  GraphQLID,
  getNullableType,
} from "graphql";

describe("Reducer", () => {
  let UserType = new GraphQLObjectType({
    name: "User",
    fields: () => ({
      id: { type: GraphQLInt },
      name: { type: GraphQLString },
      age: { type: GraphQLInt },
      friends: { type: new GraphQLList(UserType) },
    }),
  });
  let Power = new GraphQLObjectType({
    name: "Power",
    fields: {
      possible_entities: { type: UserType },
    },
  });
  let Skill = new GraphQLObjectType({
    name: "Skill",
    fields: () => ({
      id: { type: GraphQLInt },
      power: { type: GraphQLInt },
      target: { type: UserType },
    }),
  });
  let littleSchema = new GraphQLSchema({
    query: new GraphQLObjectType({
      name: "RootQueryType",
      fields: {
        users: {
          type: new GraphQLList(UserType),
          resolve() {
            return [];
          },
        },
        skills: {
          type: new GraphQLList(Skill),
          resolve() {
            return [];
          },
        },
      },
    }),
  });
  const { reducer, actions } = configure(littleSchema);
  describe("Package Data", () => {
    it("should package data correctly", () => {
      let state = reducer(
        undefined,
        actions.packageData({
          User: {
            id: 1,
            name: "Alan",
            friends: [
              {
                id: 2,
                name: "Antoine",
              },
              {
                id: 3,
                name: "Joffrey",
              },
            ],
          },
          Skill: {
            id: 1,
            power: 4,
            target: {
              id: 1,
              name: "Alan",
            },
          },
        }),
      );
      let subject = reducer(
        state,
        actions.packageData({
          User: {
            id: "2",
            friends: [{ id: 3, name: "Joffrey", age: 28 }],
          },
          Skill: [
            { id: 1, power: 5 },
            { id: 2, power: 2, target: { id: 2, name: "Antoine" } },
          ],
        }),
      );

      expect(subject).toMatchSnapshot();
    });
  });

  it("should accept only one data", () => {
    let subject = reducer(
      undefined,
      actions.packageData({
        User: {
          id: "2",
        },
      }),
    );

    expect(subject).toMatchSnapshot();
  });

  it("accepts queries", () => {
    const subject = reducer(
      undefined,
      actions.packageData(
        {
          User: {
            id: "2",
            friends: [{ id: 3, name: "Joffrey", age: 28 }],
          },
          Skill: [
            { id: 1, power: 5 },
            { id: 2, power: 2, target: { id: 2, name: "Antoine" } },
          ],
        },
        {
          request: {
            ql:
              "{ users: { id, friends { id, name, age } }, skills { id, power, target { id, name } } }",
            hash: "queryhash",
          },
          response: {
            raw: {
              users: {
                id: "2",
                friends: [{ id: 3, name: "Joffrey", age: 28 }],
              },
              skills: [
                { id: 1, power: 5 },
                { id: 2, power: 2, target: { id: 2, name: "Antoine" } },
              ],
            },
          },
        },
      ),
    );

    expect(subject).toMatchSnapshot();
  });

  describe("Remove Data", () => {
    it("should remove several data", () => {
      let subject = reducer(
        fromJS({
          entities: {
            User: {
              "1": {
                id: 1,
                name: "Alan",
              },
              "2": {
                id: 2,
                name: "Joffrey",
              },
              "3": {
                id: 3,
                name: "Antoine",
              },
            },
            Skill: {
              "1": {
                id: 1,
                power: 2,
              },
              "2": {
                id: 2,
                power: 3,
              },
            },
          },
        }),
        actions.removeData({
          User: [1, 3],
          Skill: 2,
        }),
      );

      expect(subject).toMatchSnapshot();
    });

    it("shouldn't crash if no data exists", () => {
      let subject = reducer(
        undefined,
        actions.removeData({ Users: "nonexistantid" }),
      );

      expect(subject).toMatchSnapshot();
    });
  });

  describe("Query started", () => {
    it("puts query state in reducer", () => {
      const subject = reducer(
        undefined,
        actions.queryStarted({
          request: { hash: 904239430, ql: `{ users { id } }` },
        }),
      );

      expect(subject).toMatchSnapshot();
    });
  });

  describe("Query failed", () => {
    it("marks query progress in reducer as failed", () => {
      const request = { hash: 904239430, ql: `{ users { id } }` };
      const initialState = reducer(
        undefined,
        actions.queryStarted({
          request,
        }),
      );
      const subject = reducer(
        initialState,
        actions.queryFailed({ request }, [{ error: "an error" }]),
      );

      expect(subject).toMatchSnapshot();
    });
  });
});
