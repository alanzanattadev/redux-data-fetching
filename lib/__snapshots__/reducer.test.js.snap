// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Reducer Package Data should package data correctly 1`] = `
Object {
  "entities": Object {
    "Skill": Object {
      "1": Object {
        "id": 1,
        "power": 5,
        "target": 1,
      },
      "2": Object {
        "id": 2,
        "power": 2,
        "target": 2,
      },
    },
    "User": Object {
      "1": Object {
        "age": undefined,
        "friends": Array [
          2,
          3,
        ],
        "id": 1,
        "name": "Alan",
      },
      "2": Object {
        "age": undefined,
        "friends": Array [
          3,
        ],
        "id": 2,
        "name": "Antoine",
      },
      "3": Object {
        "age": 28,
        "friends": undefined,
        "id": 3,
        "name": "Joffrey",
      },
    },
  },
  "queries": Object {},
}
`;

exports[`Reducer Remove Data should remove several data 1`] = `
Immutable.Map {
  entities: Immutable.Map {
    User: Immutable.Map {
      2: Immutable.Map {
        id: 2,
        name: "Joffrey",
      },
    },
    Skill: Immutable.Map {
      1: Immutable.Map {
        id: 1,
        power: 2,
      },
    },
  },
}
`;

exports[`Reducer Remove Data shouldn't crash if no data exists 1`] = `
Object {
  "entities": Object {},
  "queries": Object {},
}
`;

exports[`Reducer accepts queries 1`] = `
Object {
  "entities": Object {
    "Skill": Object {
      "1": Object {
        "id": 1,
        "power": 5,
        "target": undefined,
      },
      "2": Object {
        "id": 2,
        "power": 2,
        "target": 2,
      },
    },
    "User": Object {
      "2": Object {
        "age": undefined,
        "friends": Array [
          3,
        ],
        "id": 2,
        "name": "Antoine",
      },
      "3": Object {
        "age": 28,
        "friends": undefined,
        "id": 3,
        "name": "Joffrey",
      },
    },
  },
  "queries": Object {
    "queryhash": Object {
      "results": Object {
        "byEntity": Object {},
        "byQuery": Object {
          "skills": Array [
            1,
            2,
          ],
          "users": "2",
        },
      },
    },
  },
}
`;

exports[`Reducer should accept only one data 1`] = `
Object {
  "entities": Object {
    "User": Object {
      "2": Object {
        "age": undefined,
        "friends": undefined,
        "id": "2",
        "name": undefined,
      },
    },
  },
  "queries": Object {},
}
`;
